pallets/rolldown/src/lib.rs

```diff
@@ -0,0 +1,916 @@
+
+			// remove pending requests update (sequencer read))
+			PENDING_REQUESTS::<T>::mutate_exists(
+				requests_to_cancel.clone(),
+				|maybe_pending_requests_to_cancel| {
+					if let Some(ref mut pending_requests_to_cancel) =
+						maybe_pending_requests_to_cancel
+					{
+						// reduce sequencer cancel rights
+						SEQUENCER_RIGHTS::<T>::mutate_exists(canceler.clone(), |maybe_sequencer| {
+							if let Some(ref mut sequencer) = maybe_sequencer {
+								sequencer.readRights -= 1;
+							}
+						});
+
+						// create hash of pending requests
+						let hash_of_pending_request = Self::calculate_hash_of_pending_requests(
+							pending_requests_to_cancel.1.as_str(),
+						);
+
+						//get last processed request on L1
+						//deserialize json
+						let pending_requests_json =
+							// add SLASH if json is invalid to process
+							Self::deserialize_json(pending_requests_to_cancel.1.as_str()).unwrap();
+						// get last processed request on L1 and last accepted request on L1
+						let last_processed_request_on_l1 =
+							pending_requests_json.lastProccessedRequestOnL1;
+						let last_accepted_request_on_l1 =
+							pending_requests_json.lastAcceptedRequestOnL1;
+
+						// create cancel request
+						let cancel_request = Cancel {
+							updater: pending_requests_to_cancel.0.clone(),
+							canceler,
+							lastProccessedRequestOnL1: last_processed_request_on_l1,
+							lastAcceptedRequestOnL1: last_accepted_request_on_l1,
+							hash: hash_of_pending_request,
+						};
+						// increase counter for updates originating on l2
+						l2_origin_updates_counter::<T>::put(
+							Self::get_l2_origin_updates_counter() + 1,
+						);
+						// add cancel request to pending updates
+						PENDING_UPDATES::<T>::insert(
+							Self::get_l2_origin_updates_counter(),
+							Update::Cancel(cancel_request),
+						);
+						// remove whole l1l2 update (read) from pending requests
+						PENDING_REQUESTS::<T>::remove(&requests_to_cancel);
+
+						log!(info, "Pending Updates:");
+						for (request_id, update) in PENDING_REQUESTS::<T>::iter() {
+							log!(info, "request_id: {:?}:  {:?} ", request_id, update);
+						}
+						log!(info, "Pending requests:");
+						for (request_id, update) in PENDING_REQUESTS::<T>::iter() {
+							log!(info, "request_id: {:?}:  {:?} ", request_id, update);
+						}
+					} else {
+						log!(
+							info,
+							"No pending requests co cancel at dispute period end {:?}",
+							requests_to_cancel
+						);
+					}
+				},
+			);
+			Ok(().into())
+		}
+
+		// no checks if this read is correct, can put counters out of sync
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::DbWeight::get().reads_writes(1, 1).saturating_add(Weight::from_parts(40_000_000, 0)))]
+		pub fn council_update_l2_from_l1(
+			origin: OriginFor<T>,
+			input_json: String,
+		) -> DispatchResultWithPostInfo {
+			let _ = ensure_root(origin)?;
+
+			match Self::deserialize_json(&input_json) {
```

Comment:
will it be a max size of this json?

Reply:
this is same issue as max number of requests